db/skipped_sequence.go

```diff
@@ -0,0 +1,211 @@
+// the capacity of the slice to length + 100 if the current capacity is 2.5x the length
+func (s *SkippedSequenceSlice) SkippedSequenceCompact(ctx context.Context, maxWait int64) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	timeNow := time.Now().Unix()
+	indexToDelete := -1
+	for _, v := range s.list {
+		timeStamp := v.getTimestamp()
+		if (timeNow - timeStamp) >= maxWait {
+			indexToDelete++
+		} else {
+			// exit early to avoid iterating through whole slice
+			break
+		}
+	}
+	// if we have incremented the index counter, we need to the index range from the slice
+	if indexToDelete > -1 {
+		s.list = slices.Delete(s.list, 0, indexToDelete+1)
+	}
+	// resize slice to reclaim memory if we need to
+	s.clip(ctx)
+}
+
+// clip will clip the capacity of the slice to a constant value if the current capacity of the slice in 2.5x the length
+func (s *SkippedSequenceSlice) clip(ctx context.Context) {
+	// threshold is 2.5x the current length of the slice
+	threshold := 2.5 * float64(len(s.list))
+
+	if cap(s.list) > int(threshold) {
+		// check if we can safely clip without an out of bound errors
+		if (len(s.list) + ClipCapacityHeadroom) < cap(s.list) {
+			base.DebugfCtx(ctx, base.KeyCache, "clipping skipped list capacity")
+			s.list = s.list[:len(s.list):ClipCapacityHeadroom]
+		}
+	}
+}
+
+// findSequence will use binary search to search the elements in the slice for a given sequence
+func (s *SkippedSequenceSlice) findSequence(x uint64) (int, bool) {
+	i, found := slices.BinarySearchFunc(s.list, x, func(a SkippedSequenceListInterface, seq uint64) int {
+		singleSeq, ok := a.(*SingleSkippedSequence)
+		if ok {
+			if singleSeq.seq > seq {
+				return 1
+			}
+			if singleSeq.seq == seq {
+				return 0
+			}
+			return -1
+		}
+		// should never get here as it stands, will have extra handling here pending CBG-3853
+		return 1
+	})
+	return i, found
+}
+
+// removeSeq will remove a given sequence from the slice if it exists
+func (s *SkippedSequenceSlice) removeSeq(x uint64) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	index, found := s.findSequence(x)
+	if !found {
+		return fmt.Errorf("sequence %d not found in the skipped list", x)
+	}
+	// handle boarder cases where x is equal to start or end sequence on range (or just sequence for single entries)
+	if !s.list[index].isRange() {
+		// if not a range, we can just remove the single entry from the slice
+		if s.list[index].getLastSeq() == x {
+			s.list = slices.Delete(s.list, index, index+1)
+		}
+		return nil
+	}
+	// more range handling here CBG-3853, temporarily error as we shouldn't get to this point at this time
+	return fmt.Errorf("entered range handling code")
+}
+
+// insert will insert element in middle of slice maintaining order of rest of slice
+func (s *SkippedSequenceSlice) insert(index int, entry SkippedSequenceListInterface) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.list = append(s.list[:index+1], s.list[index:]...)
+	s.list[index] = entry
```

Comment:
SliceTricks recommends something a little bit different for insert:
```
s = append(s, 0 /* use the zero value of the element type */)
copy(s[i+1:], s[i:])
s[i] = x
```
Is there an advantage to the approach you've got?

Reply:
Slight difference in the benchmarking. Below is my approach currently:
```
BenchmarkInsertSkippedItem
BenchmarkInsertSkippedItem/single_entries_large_slice
BenchmarkInsertSkippedItem/single_entries_large_slice-10         	  309714	      3615 ns/op	     108 B/op	       1 allocs/op
BenchmarkInsertSkippedItem/single_entries_small_slice
BenchmarkInsertSkippedItem/single_entries_small_slice-10         	 8871403	       136.3 ns/op	     109 B/op	       1 allocs/op
```
Then with the above suggested approach:
```
BenchmarkInsertSkippedItem
BenchmarkInsertSkippedItem/single_entries_large_slice
BenchmarkInsertSkippedItem/single_entries_large_slice-10         	  311587	      3610 ns/op	     123 B/op	       2 allocs/op
BenchmarkInsertSkippedItem/single_entries_small_slice
BenchmarkInsertSkippedItem/single_entries_small_slice-10         	 7984885	       140.0 ns/op	     114 B/op	       2 allocs/op
```
Seems that the SliceTricks approach has extra allocation per op and slightly more bytes per op used (but really not much difference).

Reply:
I don't think the SliceTricks has an extra allocation, if it's implemented as:
```
	s.list = append(s.list, nil)
	copy(s.list[index+1:], s.list[index:])
	s.list[index] = entry
```
When running locally the two approaches seemed equivalent from a timing and allocation perspective.  I couldn't find anything definitive in the go docs/slice tricks about the relative benefits of:
```
	s= append(s, nil)
	copy(s[i+1:], s[i:])
versus
         s = append(s[:i+1], s.list[i:]...)
```
Now that I've spent this much time on it, though, I'm wondering whether we even expect a use case for insert, since this is an append-only slice?

Reply:
Ah I see where I went wrong now, thank you for putting me right! We need insert for when we have to split ranges when sequences arrive at the cache. So if we have range 10-20 and sequence 15 arrives, we split range into 10-14 and insert new range 16-20. I was thinking this is needed for single entries too as if we have range 10-20 and 19 arrives we can alter range to 10-18 but then replace the range and index+1 with a single entry for sequence 20.